# Core Statistics Using Python
### Hana Choi, Simon Business School, University of Rochester


# Assignment 5 Solutions


## Required packages

In [26]:
import pandas as pd
import statsmodels.formula.api as smf
from scipy import optimize

## Loading data

In [27]:
store24 = pd.read_csv("/Users/hanachoi/Dropbox/teaching/core_statistics/Assignments/Data/Store24Data.csv", thousands=',')

# Display the first few rows
print(store24.head())
print('----')

# Summary of the dataset
print(store24.describe())

   Store    Sales  Profit   MTenure    CTenure    Pop      Comp  Visibility  \
0      1  1060294  265014   0.00000  24.804930   7535  2.797888           3   
1      2  1619874  424007  86.22219   6.636550   8630  4.235555           4   
2      3  1099921  222735  23.88854   5.026694   9695  4.494666           3   
3      4  1053860  210122   0.00000   5.371663   2797  4.253946           4   
4      5  1227841  300480   3.87737   6.866530  20335  1.651364           2   

   PedCount  Res  Hours24  
0         3    1        1  
1         3    1        1  
2         3    1        1  
3         2    1        1  
4         5    0        1  
----
           Store         Sales         Profit     MTenure     CTenure  \
count  75.000000  7.500000e+01      75.000000   75.000000   75.000000   
mean   38.000000  1.205413e+06  276313.613333   45.296444   13.931499   
std    21.794495  3.045313e+05   89404.076338   57.671551   17.697517   
min     1.000000  6.993060e+05  122180.000000    0.000000   

# Questions (a) - (e)

In [28]:
# Linear regression model 
formula_input1 = 'Profit ~ MTenure + CTenure + Pop + Comp + Visibility + PedCount + Res + Hours24'
reg1 = smf.ols(formula=formula_input1, data=store24).fit()
print(reg1.summary().tables[1])

                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   7610.0415   6.68e+04      0.114      0.910   -1.26e+05    1.41e+05
MTenure      760.9927    127.086      5.988      0.000     507.258    1014.727
CTenure      944.9780    421.687      2.241      0.028     103.052    1786.904
Pop            3.6666      1.466      2.501      0.015       0.739       6.594
Comp       -2.529e+04   5491.937     -4.604      0.000   -3.63e+04   -1.43e+04
Visibility  1.263e+04   9087.620      1.389      0.169   -5518.571    3.08e+04
PedCount    3.409e+04   9073.196      3.757      0.000     1.6e+04    5.22e+04
Res         9.158e+04   3.92e+04      2.334      0.023    1.33e+04     1.7e+05
Hours24     6.323e+04   1.96e+04      3.219      0.002     2.4e+04    1.02e+05


## Question (a)

- The coefficient on MTenure (761) implies that an additional month of manager tenure is associated with an expected increase in (yearly) profit of $761.

- MTenure is statistically significant at the 1\% level (p-value is < 0.01).

## Question (b)

- Likewise, the coefficient on CTenure (945) implies that an additional month of crew tenure is associated with an expected increase in (yearly) profit of $945.

- CTenure is not statistically significant at the 1\% level (p-value is > 0.01), but it is significant at the 5% level (p-value < 0.05).

## Question (c)

- Using the coefficient on CTenure, we can infer that a 1.38-month increase in crew tenure is associated with an expected increase in (yearly) profit of 1.38*945 = $1304.1

- This is our best (point) estimate of the expected yearly increase in profit associated with a 1.38-month increase in crew tenure.

- However, there is substantial uncertainty around this estimate. In particular, a 95\% confidence interval for the increase in CTenure by 1.38-month is given by 1.38*(103.05, 1786.90) = (142.21, 2465.92), which is quite wide.

- This means with 95\% probability, we cannot rule out that the true impact of an increase of 1.38-month crew tenure on the (yearly) profit is as low as \\$142.21 and as high as \\$2465.92, a fairly wide range.

- Converting these to the quarterly profit level, we have the 95% confidence interval of 0.25*(142.21, 2465.92) = (35.55, 616.48).

- Weighing the cost and the benefit to the firm, if the average quarterly bonus for those managers increasing CTenure by 1.38 during the quarter is below \\$35.55, we can conclude that the recent store manager bonus plan is likely a successful one using a 5% significance level (if it's above \\$616.48 then not a successful one).

- If the average quarterly bonus for those managers increasing CTenure by 1.38 during the quarter is between (35.55, 616.48), we cannot conclusively determine (statistically speaking) whether the recent bonus plan is a successful or not a successful one.

- In this case, we might seek to collect more data (to increase the number of observations, or to increase the number of explanatory variables) and see if we can get a more precise estimate for the effect of CTenure on Profit.

## Question (d)

- 'Hours24' is a binary regressor taking values 0 and 1, and the coefficient estimate for 'Hours24' represents the difference in means for those open 24 hours and those not.

- The estimate implies that the average (yearly) profit is $63,233 higher for those stores open 24 hours (i.e., Hours24==1) compared to those that are not.

## Question (e)

- The 95% confidence interval for 'Pop' on profit is (0.74, 6.59).

- In words, this means that (0.74, 6.59) range will contain the true value of the parameter for 'Pop' with 95% probability.

- It also means that with 95% probability, the true impact of a unit increase in 'Pop' (i.e., an additional person living within 1/2 mile radius) on the (yearly) profit is between \\$0.74 and \\$6.59.  

## Question (f)


In [29]:
formula_input2 = 'Profit ~ MTenure + CTenure'
reg2 = smf.ols(formula=formula_input2, data=store24).fit()
print(reg2.summary().tables[1])
print('----')

print("Adjusted R2 from reg1:", reg1.rsquared_adj)
print("Adjusted R2 from reg2:", reg2.rsquared_adj)

                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    2.37e+05   1.31e+04     18.054      0.000    2.11e+05    2.63e+05
MTenure      619.8420    166.696      3.718      0.000     287.539     952.145
CTenure      810.1287    543.219      1.491      0.140    -272.758    1893.015
----
Adjusted R2 from reg1: 0.5940158638411897
Adjusted R2 from reg2: 0.19504413998277492


- The site location factors are important for 1) controlling for possible omitted variable bias and 2) improving the precision of our estimates.

- Omitted variable bias might arise if Store 24 is taking steps to ensure that better (more experienced) managers and crews are assigned to particular types of stores (could be the better locations, could be the worse ones, we can't be sure). 

- Dropping the site location factors from the analysis (see reg2 above), we find that the adjusted R^2 falls from .59 to .20, implying that site location factors are quite important in explaining the variation in profits. 

- Not surprisingly, the t-stats on MTenure and CTenure also fall, as there is now more unexplained variation adding noise to our estimates.

- Finally, the coefficient estimates on MTenure and CTenure also fall, suggesting that omitted variable bias is indeed a concern (were we to leave the site location factors out). They should definitely be included.

## Question (g)

In [30]:
formula_input3 = 'Profit ~ MTenure + I(MTenure**2) + CTenure + I(CTenure**2) + Pop + Comp + Visibility + PedCount + Res + Hours24'
reg3 = smf.ols(formula=formula_input3, data=store24).fit()
print(reg3.summary().tables[1])
print('----')

print("Adjusted R2 from reg1:", reg1.rsquared_adj)
print("Adjusted R2 from reg3:", reg3.rsquared_adj)

                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept       -2.418e+04   6.23e+04     -0.388      0.699   -1.49e+05       1e+05
MTenure          1732.5482    312.207      5.549      0.000    1108.842    2356.254
I(MTenure ** 2)    -4.8302      1.443     -3.348      0.001      -7.712      -1.948
CTenure          2460.3589    974.711      2.524      0.014     513.151    4407.567
I(CTenure ** 2)   -16.9008     10.376     -1.629      0.108     -37.629       3.828
Pop                 3.4752      1.397      2.487      0.015       0.684       6.267
Comp            -2.546e+04   5136.143     -4.956      0.000   -3.57e+04   -1.52e+04
Visibility       1.976e+04   8750.212      2.258      0.027    2281.114    3.72e+04
PedCount         3.914e+04   8486.859      4.612      0.000    2.22e+04    5.61e+04
Res              6.271e+04   3.86e+04      1.626      0.109   -1.43e+04     

- We might expect there to be diminishing returns to tenure for both managers and crews since, at some point, they have presumably learned all that they can. We can account for this possibility using a non-linear specification. 

- Since part of our goal here is to learn whether a non-linear specification is appropriate, I will use a quadratic regression approach rather than taking logs of MTenure and CTenure. 

- Alternatively, you could have taken logs of MTenure and CTenure, but this would have required you to assume that diminishing returns exist, rather than allowing for either diminishing, increasing, or constant returns. Moreover, some of the values of MTenure are zero, for which the log function is not defined.

- This requires creating two new variables (MTenure^2 and CTenure^2), which are the squares of the original tenure variables, and including them in the analysis. Doing so, we find that both manager and crew tenure display diminishing returns (the coefficients on MTenure^2 and Ctenure^2 are negative), although only MTenure^2 is significant at the 5% or 1% level.

- We also find that the adjusted R^2 increases from .59 to .66.

## Question (h)

- Here we need to find the point at which we reach the top of the hill implied by the quadratic relationship between manager tenure and profit. 

- To do so, I used 'minimize_scalar' to maximize Profit (here, just the component of profit associated with MTenure), by changing the value of MTenure.

$$1732.548 * MTenure - 4.830 * (MTenure^2)$$

- We find that the manager's impact on profitability begins to diminish after 179 months (about 15 years).

- You could also have done this by setting the derivative equal to zero and solving for the corresponding (optimal) level of tenure.

In [31]:
# Define the profit function
def profit(mtenure, a, b):
    profit = a * mtenure + b * mtenure**2
    return profit

# Maximize profit by minimizing the negative profit
a = reg3.params['MTenure']
b = reg3.params['I(MTenure ** 2)']
result = optimize.minimize_scalar(lambda mtenure: -profit(mtenure, a, b), bounds=(0, 1000), method='bounded')
print("Optimal MTenure:", result.x)
print("Profit at the Optimal MTenure:", -result.fun)

Optimal MTenure: 179.3435313308041
Profit at the Optimal MTenure: 155360.65184570255
